training/components.py

```diff
@@ -292,154 +230,175 @@ def list_phase1_final_model():
+        models.V1EnvVar(name="XDG_CACHE_HOME", value="/tmp"),
+        models.V1EnvVar(name="TRITON_CACHE_DIR", value="/tmp"),
+        models.V1EnvVar(name="HF_HOME", value="/tmp"),
+        models.V1EnvVar(name="TRANSFORMERS_CACHE", value="/tmp"),
+    ]
+
+    # Get master and worker container specs
+    master_container_spec = kfto_utils.get_container_spec(
+        base_image=base_image,
+        name="pytorch",
+        resources=resources_per_worker,
+        volume_mounts=volume_mounts_master,
+    )
+
+    # In the next release of kubeflow-training, the command
+    # and the args will be a part of kfto_utils.get_container_spec function
+    master_container_spec.command = command
+    master_container_spec.args = master_args
+
+    master_container_spec.env = env_vars
+
+    worker_container_spec = kfto_utils.get_container_spec(
+        base_image=base_image,
+        name="pytorch",
+        resources=resources_per_worker,
+        volume_mounts=volume_mounts_worker,
+    )
+    worker_container_spec.command = command
+    worker_container_spec.args = worker_args
+    worker_container_spec.env = env_vars
+
+    # create master pod spec
+    master_pod_template_spec = kfto_utils.get_pod_template_spec(
+        containers=[master_container_spec],
+        volumes=volumes,
+    )
+
+    # create worker pod spec
+    worker_pod_template_spec = kfto_utils.get_pod_template_spec(
+        containers=[worker_container_spec],
+        volumes=volumes,
+    )
+
+    logging.getLogger(__name__).setLevel(logging.INFO)
+    logging.info("Generating job template.")
+    logging.info("Creating TrainingClient.")
+
+    # Initialize training client
+    # This also finds the namespace from /var/run/secrets/kubernetes.io/serviceaccount/namespace
+    # And it also loads the kube config
+    training_client = TrainingClient()
+    namespace = training_client.namespace
+
+    # Create pytorch job spec
+    job_template = kfto_utils.get_pytorchjob_template(
+        name=name,
+        namespace=namespace,
+        worker_pod_template_spec=worker_pod_template_spec,
+        master_pod_template_spec=master_pod_template_spec,
+        num_workers=nnodes,
+        num_procs_per_worker=nproc_per_node,
+    )
+    # Save the pytorch job yaml for record keeping and debugging
+    with open(pytorchjob_output_yaml.path, "w", encoding="utf-8") as f:
+        f.write(job_template.to_str(), f)
+
+    # Run the pytorch job
+    logging.info(f"Creating PyTorchJob in namespace: {namespace}")
+    training_client.create_job(job_template, namespace=namespace)
+
+    expected_conditions = ["Succeeded", "Failed"]
+    logging.info(f"Monitoring job until status is any of {expected_conditions}.")
+
+    def wait_for_job_get_logs(
+        name: str,
+        namespace: str = None,
+        job_kind: str = None,
+        expected_conditions: list = ["Succeeded"],
+        wait_timeout: int = 600,
+        polling_interval: int = 15,
+        timeout: int = 1000,
```

Comment:
Can you please explain this function to me a bit? I don't follow.. Can't be a simple call:

```
training_client.get_jobs_log(follow=True, ...)
training_client.get_job_conditions()
```

Based on the [`get_jobs_log()` sources](https://github.com/kubeflow/training-operator/blob/374e88af2d28d49e1c40008238f6e039b5677177/sdk/python/kubeflow/training/api/training_client.py#L1210-L1255) the `follow=True` should be blocking and stream all the logs until all the pods finish.

Is there anything wrong with that approach? I may have overlooked it...

Reply:
I tried `follow=True` flag but it doesn't block the launcher container from finishing. I also tried [wait_for_job_condition](https://github.com/kubeflow/training-operator/blob/374e88af2d28d49e1c40008238f6e039b5677177/sdk/python/kubeflow/training/api/training_client.py#L888) which blocks the launcher but doesn't stream logs. If I use both of them, the `training_client.get_jobs_log(follow=True, ...)` prints the log stream to stdout but if we want to see it in pod logs and on the UI logs for the launcher, we have to use `logging.info()` to show them. 

The above function is similar to `wait_for_job_condition` but adds the get pod logs while waiting feature. It's good that you bring it up, maybe there can be a better way to do all of this.

Reply:
Just curious.. this you also try the `wait_for_job_condition` with `callback` param?

https://github.com/kubeflow/training-operator/blob/374e88af2d28d49e1c40008238f6e039b5677177/sdk/python/kubeflow/training/api/training_client.py#L896

It should allow us to get the logs via a callback function on each polling interval.

Reply:
I did not try that. Turns out, the callback function option with the get_job_logs(..., follow=True) does work for us. The only caveat is that the callback function can only have the job object as input whereas get_job_logs takes in the job name as an input. So, as a workaround I had to create another function to bridge that mismatch. I'll add that shortly, but awesome suggestion, it significantly reduces the unnecessary code 💯